From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Biquaternions <biquaternions@serlith.net>
Date: Sun, 3 Aug 2025 20:38:56 -0500
Subject: [PATCH] Fish Schedule Async Tasks For End Of Tick

This implementation allows async threads to still do async operations as long as the world is not ticking.
If the world ticking, the "request" will be scheduled to be ran sync at the end of the tick.
  * Not doing this, means that if the world is always ticking (mspt > 50) there's a chance that the async thread will sleep indefinitely.
If the world is not ticking, the "request" will be performed async by the calling thread.
If the world is about to tick, but the async "request" is not done yet, the world will wait until its done and every new "request" will be scheduled.
This scheduling logic is very performant as long as none of the plugins is spamming world/block requests.

diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 9d5c20282f88ed30db16cd6757e1bd3877957612..f369185156820a902fb3530c2ae6165bec77a695 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -842,7 +842,15 @@ public class ServerLevel extends Level implements ServerEntityGetter, WorldGenLe
         profilerFiller.push("entityManagement");
         // Paper - rewrite chunk system
         profilerFiller.pop();
+        // Fish start - Schedule tasks for end of tick
+        Runnable task;
+        while ((task = this._fish_endOfTickTasks.poll()) != null) {
+            task.run();
+        }
+        // Fish end - Schedule tasks for end of tick
     }
+    public final java.util.Queue<Runnable> _fish_endOfTickTasks = new java.util.concurrent.ConcurrentLinkedQueue<>(); // Fish - Schedule tasks for end of tick
+    public final java.util.concurrent.locks.ReentrantReadWriteLock lock = new java.util.concurrent.locks.ReentrantReadWriteLock(true);
 
     @Override
     public boolean shouldTickBlocksAt(long chunkPos) {
